Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -0,0 +1,379 @@
+		{
+			// This swaps Role/RemoteRole as we write it
+			const FRepLayoutCmd& SwappedCmd = (!bIsServer && Parent.RoleSwapIndex != -1) ? Cmds[Parents[Parent.RoleSwapIndex].CmdStart] : Cmd;
+
+			uint8* Data = (uint8*)Object + SwappedCmd.Offset;
+
+			if (bIsInitialData || GetPropertyCount(ComponentObject, FieldId, Cmd.Property) > 0 || ClearedIds->Find(FieldId) != INDEX_NONE)
+			{
+				if (Cmd.Type == REPCMD_DynamicArray)
+				{
+					ApplyArray(ComponentObject, FieldId, Cast<UArrayProperty>(Cmd.Property), Data, SwappedCmd.Offset, Cmd.ParentIndex);
+				}
+				else
+				{
+					ApplyProperty(ComponentObject, FieldId, 0, Cmd.Property, Data, SwappedCmd.Offset, Cmd.ParentIndex);
+				}
+
+				if (Cmd.Property->GetFName() == NAME_RemoteRole)
+				{
+					// Downgrade role from AutonomousProxy to SimulatedProxy if we aren't authoritative over
+					// the client RPCs component.
+					UByteProperty* ByteProperty = Cast<UByteProperty>(Cmd.Property);
+					if (!bIsServer && !bAutonomousProxy && ByteProperty->GetPropertyValue(Data) == ROLE_AutonomousProxy)
+					{
+						ByteProperty->SetPropertyValue(Data, ROLE_SimulatedProxy);
+					}
+				}
+
+				// Parent.Property is the "root" replicated property, e.g. if a struct property was flattened
+				if (Parent.Property->HasAnyPropertyFlags(CPF_RepNotify))
+				{
+					if (Parent.RepNotifyCondition == REPNOTIFY_Always || !Cmd.Property->Identical(RepState->StaticBuffer.GetData() + SwappedCmd.Offset, Data))
+					{
+						RepNotifies.AddUnique(Parent.Property);
+					}
+				}
+			}
+		}
+	}
+
+	Channel->PostReceiveSpatialUpdate(Object, RepNotifies);
+}
+
+void ComponentReader::ApplyProperty(Schema_Object* Object, Schema_FieldId Id, std::uint32_t Index, UProperty* Property, uint8* Data, int32 Offset, uint16 ParentIndex)
+{
+	if (UStructProperty* StructProperty = Cast<UStructProperty>(Property))
+	{
+		TArray<uint8> ValueData = Schema_IndexPayload(Object, Id, Index);
+		// A bit hacky, we should probably include the number of bits with the data instead.
+		int64 CountBits = ValueData.Num() * 8;
+		TSet<UnrealObjectRef> NewUnresolvedRefs;
+		FSpatialNetBitReader ValueDataReader(PackageMap, ValueData.GetData(), CountBits, NewUnresolvedRefs);
+		bool bHasUnmapped = false;
+
+		ReadStructProperty(ValueDataReader, StructProperty, NetDriver, Data, bHasUnmapped);
+
+		if (bHasUnmapped)
+		{
+			ObjectReferencesMap.Add(Offset, FObjectReferences(ValueData, CountBits, NewUnresolvedRefs, ParentIndex, Property));
+			UnresolvedRefs.Append(NewUnresolvedRefs);
+		}
+		else if (ObjectReferencesMap.Find(Offset))
+		{
+			ObjectReferencesMap.Remove(Offset);
+		}
+	}
+	else if (UBoolProperty* BoolProperty = Cast<UBoolProperty>(Property))
+	{
+		BoolProperty->SetPropertyValue(Data, Schema_IndexBool(Object, Id, Index) != 0);
+	}
+	else if (UFloatProperty* FloatProperty = Cast<UFloatProperty>(Property))
+	{
+		FloatProperty->SetPropertyValue(Data, Schema_IndexFloat(Object, Id, Index));
```

Comment:
Note that this behaviour is not completely identical to `Schema_GetFloat` with this code (when `Index == 0`). `Schema_GetFloat` will always return the _last_ element rather than the first one.

I think this is fine as it is though though, but it might instead be worth passing a `bool` into `ApplyProperty`  to control whether this is an array or not, and either use `Schema_GetFloat` and friends, or use `Schema_GetFloatCount` as the index.

Reply:
It's always the server worker populating the component updates, so I don't see how `Schema_GetFloat` is ever more correct than `Schema_IndexFloat` with index 0.
If there's a mistake when populating the update, then why would the last added float be better than the first added one?

Reply:
Yeah, it probably doesn't make a different. I just wanted to highlight that the semantics are not completely identical, and whether that is acceptable (which it is).